docs/DafnyRef/UserGuide.md

```diff
@@ -142,7 +142,7 @@ The command-line also expects the following:
 working directory. A command-line argument not matching a known option is considered a filepath, and likely one
 with an unsupported suffix, provoking an error message.
 - Files containing dafny code must have a `.dfy` suffix.
-- There must be at least one `.dfy` file.
+- There must be at least one `.dfy` file (except in the case of `dafny format`, see the [Dafny format section](#sec-formatting)) 
```

Comment:
I strongly prefer that dafny format not be an exception. The argument could be a directory, or an option like --repo. But an empty argument list is often an accident (e.g. dafny format *.dy), and this one would start a lot of work that might modify files.

Reply:
Your example is not a problem for the current dafny formatter, it reports as expected

```
> ./Binaries/Dafny format *.dy
*** Error: 'C:\Users\mimayere\Documents\dafny\*.dy': Filename extension '.dy' is
 not supported. Input files must be Dafny programs (.dfy) or supported auxiliary
 files ()

I also tried with `unexisting*.dfy` and it throws an exception (might not be good but at least it's complaining it cannot read the file)

So here are my thoughts
* I think there are two schools for formatters.
* Some formatting tools I got inspiration from do the same:
  * [dotnet format](https://github.com/dotnet/format) formats all .cs files in the current directory
  * [scalafmt](https://scala-cli.virtuslab.org/docs/commands/fmt/) has the same behavior, it formats all Scala files
  * [cargo fmt](https://github.com/rust-lang/rustfmt) does the same for rust
* Some formatting tools do your suggested approach and don't support zero arguments
  * [go fmt <path>](https://pkg.go.dev/cmd/gofmt), without path it process the stdin
  * [tsfmt -r](https://www.npmjs.com/package/typescript-formatter) does recursive formatting only if there is a project file
  * [JIndent](http://www.newforms-tech.com/products/jindent/components/console) requires an explicit directory.

I remember you worked a lot in Java before, so that's how I understand your strong opinion. I come from a Scala background so I'm used to the first option and I never had an issue as you suggested. At worst, I could kill the formatter.
I personally prefer the first approach because it's a simple action to do - no big deal to format files anyway, if people used the formatter before, their files should be already formatted using it.
I'm ok adding support for directory names as well, down the road. For this first version, I prefer to stick to simplicity. Files => format one file. Nothing => format the directory recursively.

Reply:
Ah, that behavior (in bash anyway) depends on whether `nullglob` is set. So some users will see `*.dy` stay `*.dy` (if there are no .dy files) and some will see it be replaced by an empty list.

So I'm still strongly for requiring a specific argument (though allowing it to be stdin if there is no argument is OK).
I think it much safer as the command is altering files (recursively!) in place. 

The problem with your 'simplicity' is that it would be a breaking change to then require a directory in the future. If you can currently do the current directory recursively, it seems a very simple change to require a directory and format an arbitrary directory instead of the current directory. And writing `dafny format .` is hardly harder and much clearer than `dafny format`